[ BB2-897 ] Nightly global state metrics additional counts

dtisza1 opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
[BB2-897](https://jira.cms.gov/browse/BB2-897)


**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
Update the nightly global-state type logging to include grant & archived grant bene counts (for more accurate counts) and a few other per application log items.


Summary:

The following types of counts are added to the nightly global metrics logging:

NOTE: A "*" in the naming can be either "real" or "synthetic".

See the confluence document for details about the items (in last section):  https://confluence.cms.gov/display/BB2/BFD-Insights+Metrics+Review

* TOP LEVEL:
  - crosswalk_*_bene_count
  - crosswalk_table_count
  - crosswalk_archived_table_count
  - grant_*_bene_count
  - grant_table_count
  - grant_archived_table_count
  - grant_*_bene_deduped_count
  - grantarchived_*_bene_deduped_count
  - grant_and_archived_*_bene_deduped_count
  - token_*_bene_deduped_count
  - token_table_count
  - token_archived_table_count
  - existing metric: *_bene_cnt

* PER APPLICATION:
  - grant_*_bene_count
  - grant_table_count
  - grant_archived_table_count
  - grantarchived_*_bene_deduped_count
  - grant_and_archived_*_bene_deduped_count
  - token_*_bene_count
  - token_table_count
  - token_archived_table_count
  - existing metric:  real/synth _bene_cnt


### What Does This PR Do?

<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->
**The following is a list of changes from the commits (including additional comments):**

NOTE: The "black" formatting utility was run against all files worked on.

* Create crosswalk counts function
  - Create `get_crosswalk_bene_counts()` function that returns top level Crosswalk table metrics. Located in apps/dot_ext/models.py

* Create grant counts function
  - Create `get_grant_bene_counts(application)` function that returns top level DataAccessGrant and ArchivedDataAccessGrant table metrics. If an `application` is passed in, it returns counts sprecific to that application. Otherwise the counts are top level.  Located in apps/authorization/models.py

* Create toekn counts function
  - Create `get_token_bene_counts(application=app)` function that returns top level AccessToken and ArchivedAccessToken table metrics. If an `application` is passed in, it returns counts sprecific to that application. Otherwise the counts are top level.  Located in apps/dot_ext/models.py

* Add elapsed time to crosswalk, grant and token functions for query timing info.

* Create global state metrics logging unit tests
  - Update JSON SCHEMA for GLOBAL_STATE_METRICS_LOG_SCHEMA and GLOBAL_STATE_METRICS_PER_APP_LOG_SCHEMA in apps/logging/tests/audit_logger_schemas.py
  - Create a helper method _create_range_users_app_token_grant(self, start_fhir_id, count, app_name) hat creates a RANGE of users connected to an application with Crosswalk, Access Token, and Grant for use in tests.  Returns a DICT of users granted by fhir_id.  For the BaseApiTest class in apps/test.py
  - Create tests to validate log metrics items after creating/removing a variety of real/synth users, apps and grants using BaseApiTest helper function.  This updates the test_management_command_logging() test method in apps/logging/tests/test_loggers_management_command.py

* Update count queries to handle NULL/empty crosswalks in queries in grant and token functions.
  - NOTE:  This handles cross referenced Null values in the queries. For example, if there is a grant, but not a corresponding crosswalk record.
  - Add test number 5 for "Test that crosswalk record with fhir_id = "" is not counted." in `TestLoggersGlobalMetricsManagementCommand.test_management_command_logging()` method in apps/logging/tests/test_loggers_management_command.py
  - Add test number 6 for "Test that missing/NULL crosswalk record is not counted" in `TestLoggersGlobalMetricsManagementCommand.test_management_command_logging()` method in apps/logging/tests/test_loggers_management_command.py

* Refactoring of functions and namings per the following confluence document:  https://confluence.cms.gov/display/BB2/BFD-Insights+Metrics+Review

* Add table total record counts to crosswalk, grant and token functions and logging.
  - Add items crosswalk_table_count & crosswalk_archived_table_count
  - Add items grant_table_count & grant_archived_table_count
  - Add items token_table_count & token_archived_table_count
  - Update log schema for new logging items in apps/logging/tests/audit_logger_schemas.py
  - Add tests for new logging items in apps/logging/tests/test_loggers_management_command.py
  - Refactor validate methods to be simpler by iterating over a fields list for _validate_global_state_metrics_log() and _validate_global_state_per_app_metrics_log() in apps/logging/tests/test_loggers_management_command.py


### What Should Reviewers Watch For?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply for some reason.

Common items include:
* Is this likely to address the goals expressed in the user story?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is user input properly sanitized & handled?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
-->
If you're reviewing this PR, please check these things, in particular:

* Are there any unhandled and/or untested edge cases you can think of?
* Does this make any backwards-incompatible changes that might break end user clients?
* Can you find any bugs if you run the code locally and test it manually?
* Naming in general? Do the namings of variables, methods, classes, and other items make sense? Are they too verbose, or not verbose enough? Any redundancies? Are the comments descriptive enough?
* Are audit logging items properly handled? Is sensitive information either not included or hashed?


There are two ways to run the management command:

* Via tests:

```
$ clear; docker-compose exec web python runtests.py apps.logging.tests.test_loggers_management_command
```

* Via docker-compose and manage.py:

```
$ clear; docker-compose exec web python manage.py log_global_state_metrics
```


### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.


### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->



### Submitter Checklist

<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [ X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [ X] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [X ] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X ] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X ] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X ] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X ] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

dtisza1 commented on apps/authorization/models.py:
```diff
@@ -70,11 +73,217 @@ def update_grants(*args, **kwargs):
         "grants": grant_count,
     }
+
+
```
START-CODE-FOR-REVIEW-SECTION

Creates `get_grant_counts()` and `get_grant_by_app_counts(application)` functions that returns `DataAccessGrant` and `ArchivedDataAccessGrant` related model counts.

dtisza1 commented on apps/authorization/models.py:
```diff
@@ -70,11 +73,217 @@ def update_grants(*args, **kwargs):
+        "synth_bene_arch_distinct_cnt": synth_bene_arch_distinct_cnt,
+        "real_bene_union_cnt": real_bene_union_cnt,
+        "synth_bene_union_cnt": synth_bene_union_cnt,
+    }
```
END-CODE-FOR-REVIEW-SECTION

dtisza1 commented on apps/dot_ext/models.py:
```diff
@@ -343,4 +403,80 @@ def get_application_require_demographic_scopes_count():
 
 
-post_delete.connect(archive_token, sender='oauth2_provider.AccessToken')
+def get_access_token_counts():
```
START-CODE-FOR-REVIEW-SECTION

Creates `get_access_token_counts()` and `get_access_token_by_app_counts(application)` functions that return access_token related counts.

dtisza1 commented on apps/dot_ext/models.py:
```diff
@@ -343,4 +403,80 @@ def get_application_require_demographic_scopes_count():
+        "real_bene_distinct_cnt": real_bene_distinct_cnt,
+        "synth_bene_distinct_cnt": synth_bene_distinct_cnt,
+    }
+
```
END-CODE-FOR-REVIEW-SECTION

dtisza1 commented on apps/fhir/bluebutton/models.py:
```diff
@@ -224,28 +253,35 @@ def __init__(self, req_response=Response):
 
 
 def check_crosswalks():
+    start_time = datetime.utcnow().timestamp()
```
START-CODE-FOR-REVIEW-SECTION

Adds an elapsed time around Crosswalk queries.

dtisza1 commented on apps/fhir/bluebutton/models.py:
```diff
@@ -224,28 +253,35 @@ def __init__(self, req_response=Response):
         "synthetic": synth_count,
         "real": real_count,
+        "elapsed": elapsed_time,
     }
```
END-CODE-FOR-REVIEW-SECTION

dtisza1 commented on apps/fhir/bluebutton/tests/test_models.py:
```diff
@@ -138,7 +138,9 @@ def test_crosswalk_real_synth_query_managers(self):
                               user_mbi_hash='987654321aaaa11111aaaa195a47a82a2cd6f46e911660fe9775f6e00000000' + str(cnt))
 
-        self.assertEqual("{'synthetic': 7, 'real': 5}", str(check_crosswalks()))
+        cc = check_crosswalks()
```
START-CODE-FOR-REVIEW-SECTION

Update test to handle change to function return.

dtisza1 commented on apps/fhir/bluebutton/tests/test_models.py:
```diff
@@ -138,7 +138,9 @@ def test_crosswalk_real_synth_query_managers(self):
+        cc = check_crosswalks()
+        self.assertEqual(cc["synthetic"], 7)
+        self.assertEqual(cc["real"], 5)
 
```
END-CODE-FOR-REVIEW-SECTION

dtisza1 commented on apps/logging/loggers.py:
```diff
@@ -1,11 +1,22 @@
-import apps.logging.request_logger as logging
-
+from datetime import datetime
```
START-CODE-FOR-REVIEW-SECTION

Adds logging items that are returned from `check_crosswalks()`,  `get_access_token_counts()` and  `get_grant_counts()` functions.

dtisza1 commented on apps/logging/loggers.py:
```diff
@@ -51,40 +122,60 @@ def log_global_state_metrics(group_timestamp=None):
+        "---    Applications (global_state_metrics_per_app) metrics elapsed time:  ",
+        elapsed_time,
+    )
+    print("---")
```
END-CODE-FOR-REVIEW-SECTION

dtisza1 commented on apps/logging/management/commands/log_global_state_metrics.py:
```diff
@@ -2,12 +2,16 @@
-        # Timestamp used to group multiple logging events from this command
-        group_timestamp = datetime.now(timezone.utc).astimezone().replace(microsecond=0).isoformat()
+        # Timestamp used to group both types of logging events from this command
+        group_timestamp = format_timestamp(datetime.now())
```
Utilize `format_timestamp()` utility function here.

dtisza1 commented on apps/logging/tests/audit_logger_schemas.py:
```diff
@@ -352,40 +554,89 @@ def get_pre_fetch_fhir_log_entry_schema(version):
   Log entry schema used for tests in apps/logging/tests/test_loggers_management_command.py
-'''
+"""
 GLOBAL_STATE_METRICS_LOG_SCHEMA = {
```
START-CODE-FOR-REVIEW-SECTION

In this file, only the `GLOBAL_STATE_METRICS_LOG_SCHEMA` and `GLOBAL_STATE_METRICS_PER_APP_LOG_SCHEMA` JSON schema were updated. These are used to validate JSON logging.

dtisza1 commented on apps/logging/tests/audit_logger_schemas.py:
```diff
@@ -394,14 +645,37 @@ def get_pre_fetch_fhir_log_entry_schema(version):
+        "user_date_joined",
+        "user_last_login",
+        "user_organization",
+    ],
```
END-CODE-FOR-REVIEW-SECTION

dtisza1 commented on apps/logging/tests/test_loggers_management_command.py:
```diff
@@ -1,20 +1,37 @@
+import copy
```
START-CODE-FOR-REVIEW-SECTION

This adds tests to validate metric results that are logged.

There are 4x sections of testing. See the comment labels for each section for info about what is being tested:  TEST #1 - 4

dtisza1 commented on apps/logging/tests/test_loggers_management_command.py:
```diff
@@ -35,98 +52,448 @@ def _validateJsonSchema(self, schema, content):
-            # Validate with schema
-            self.assertTrue(self._validateJsonSchema(GLOBAL_STATE_METRICS_PER_APP_LOG_SCHEMA, log_dict))
-            cnt = cnt + 1
+        self._validate_global_state_per_app_metrics_log(validate_apps_dict)
```
END-CODE-FOR-REVIEW-SECTION

dtisza1 commented on apps/test.py:
```diff
@@ -286,34 +322,63 @@ def _create_user_app_token_grant(self, first_name, last_name, fhir_id, app_name,
 
         return user, application, access_token
+
+    def _create_range_users_app_token_grant(self, start_fhir_id, count, app_name):
```
START-CODE-FOR-REVIEW-SECTION

Created a helper method for use in test code.

dtisza1 reviewed (commented):
The changes in this PR contain a lot of Black formatting tool changes.

To make it easier to read the main task related  changes, I've added the following comment labels to the start/end of code sections:  START-CODE-FOR-REVIEW-SECTION & END-CODE-FOR-REVIEW-SECTION

Be sure to expand the collapsed files to find these in files with a large number of changes.

Thank you!

nbragdon commented on apps/authorization/models.py:
```diff
@@ -70,11 +73,217 @@ def update_grants(*args, **kwargs):
+    start_time = datetime.utcnow().timestamp()
+
+    real_bene_distinct_cnt = (
+        DataAccessGrant.objects.order_by("beneficiary__username")
```
Have you done any perf testing on these queries?

dtisza1 replied:
@nbragdon Not yet. Planning to start testing this branch in TEST, SANDBOX, PROD via canaries soon (today).

I've been reviewing metrics in NewRelic for the 3AM time frame when this nightly job is currently running. It doesn't impact the system very much per the Postgres related metrics.  

Also, the Jenkins job has been taking about 2-minutes to complete, but this includes the container setup and other stuff before the Django command is run.

This PR includes timings for the queries, so we can get more information. Currently we can't tell from the Jenkins logs.

Will report findings from testing to review before we can approve & merge this in.

nbragdon commented on apps/logging/loggers.py:
```diff
@@ -15,34 +26,94 @@
+        "type": "global_state_metrics",
+        "group_timestamp": group_timestamp,
+        "real_bene_cnt": crosswalk_counts.get("real", None),
+        "synth_bene_cnt": crosswalk_counts.get("synthetic", None),
```
Is the synthetic bene count... the number of synthetic bene's currently being used?

dtisza1 replied:
@nbragdon Those two counts are from the Crosswalk table.  This is our largest type of count. It would include all benes that had a successful medicare-gov login. For those that have successfully authenticated, we need the Grant related counts (a subset of the crosswalk counts).

nbragdon replied:
I was just curious why we wanted to calculate the synthetic bene count... since we already know the total number of synthetic bene's (40k). What fine grained information is this giving us?

dtisza1 replied:
> I was just curious why we wanted to calculate the synthetic bene count... since we already know the total number of synthetic bene's (40k). What fine grained information is this giving us?

@nbragdon The 40K is the total synth benes that are available to BB2 via SLSx/Medicare.gov.  This counts the number of them that have connected or being used in BB2.

For example, when running this last time in Sandbox, the `synth_bene_cnt` (in the Crosswalk table) was 14,633.

nbragdon commented on apps/authorization/models.py:
```diff
@@ -70,11 +73,217 @@ def update_grants(*args, **kwargs):
+    }
+
+
+def get_grant_by_app_counts(application):
```
Maybe we could change these function names to be a bit more descriptive of what data we are trying to get from performing these operations?

nbragdon commented on apps/logging/loggers.py:
```diff
@@ -51,40 +122,60 @@ def log_global_state_metrics(group_timestamp=None):
 
+    elapsed_time = round(datetime.utcnow().timestamp() - start_time, 3)
+
     print("---")
```
Do we need these print lines? Will this make parsing this harder?

dtisza1 replied:
@nbragdon These are just for the purpose of showing the activity that occurred when reviewing the Jenkins console logging.

For example, here is part of the console log for a last run in Sandbox. I replaced some of the numbers with XXXXX.

```

TASK [Display command output: log_global_state_metrics] ************************
ok: [10.246.44.75] => {
    "msg": [
        "---",
        "---RUNNING DJANGO COMMAND:  log_global_state_metrics",
        "---",
        "---",
        "---    Wrote top level log entry:  {'type': 'global_state_metrics', 'group_timestamp': '2021-11-05T07:01:49+00:00', 'real_bene_cnt': 0, 'synth_bene_cnt': 14632, 'global_apps_active_cnt': XXXXX, 'global_apps_inactive_cnt': 0, 'global_apps_require_demographic_scopes_cnt': XXXXX}",
        "---",
        "---",
        "---    Wrote per application log entries:  XXXXX",
        "---",
        "SUCCESS",
        ""
    ]
}

```

nbragdon commented on apps/logging/utils.py:
```diff
@@ -0,0 +1,14 @@
+
+def format_timestamp(dt):
+    """
+    Returns an ISO 6801 format string in UTC that works well with AWS Glue/Athena
```
Nice comment 👍

JFU-GIT reviewed (approved):
checked out on local and worked as expected, it might be good to deploy to a canary to collect some performance data.

dtisza1 commented:
I just performed testing in a Sandbox canary.

The Django command took about 85-seconds to complete.  The top level queries took 0.388 seconds and per application queries took 83.4 seconds.

During that short time frame throughput, database, error rates, and other metrics looks good via NewRelic.

I believe this is looking good and ready to merge in.

nbragdon reviewed (approved):
This is looking good, I'll leave the function naming decision up to you.

JFU-GIT commented on apps/authorization/models.py:
```diff
@@ -70,11 +73,235 @@ def update_grants(*args, **kwargs):
+    # Grant real/synth bene counts (includes granted to multiple apps)
+    start_time = datetime.utcnow().timestamp()
+
+    real_bene_cnt = DataAccessGrant.objects.filter(
```
would count on crosswalk directly give the same result?

    real_bene_distinct_cnt = (
        Crosswalk.objects.filter(~Q(fhir_id__startswith="-")).count()
    )

    synth_bene_distinct_cnt = (
        Crosswalk.objects.filter(Q(fhir_id__startswith="-")).count()
    )

no need to check fhir for null and empty, they are enforced by db constrain?

dtisza1 replied:
@JFU-GIT  For grant and token aggregates with the Crosswalk, I was getting some slightly invalid counts in Sandbox until adding the "_isnull=False" to the queries.  I can't recall, but I think related to the Crosswalk default setting of "default=None".

From the constraints, there can only be one with FHIR_ID == "" so we want to skip that one, if it exists.

JFU-GIT reviewed (approved):
feels that some aggregate could be optimized (see comments on code lines), otherwise, looks good!

dtisza1 commented:
@nbragdon @JFU-GIT  I believe this PR is ready and working well!

I'm re-requesting reviews, since there have been changes since your approvals:

- Refactored namings.
- Refactored functions to be simpler and easier to read, especially with the tests.
- Added some more tests.
- Added some additional full table record counts.  Even though these can be derived form other metrics, they are a good sanity check. 
- Updated documentation on metric values in confluence.
- Performed testing via canaries in both Sandbox and PROD.

Below are timings from running via the canaries just now:

Sandbox:
---    Top level (global_state_metrics) metrics elapsed time:   0.83
---    Applications (global_state_metrics_per_app) metrics elapsed time:   84.651

PROD:
---    Top level (global_state_metrics) metrics elapsed time:   16.168
---    Applications (global_state_metrics_per_app) metrics elapsed time:   22.875

JFU-GIT reviewed (approved):
looks good to me

oragame reviewed (approved):
This looks good to me! Nice work on the confluence docs also
